packages/blockchain-wallet-v4-frontend/src/modals/BuySell/AddCardVgs/AddCardVgs.tsx

```diff
@@ -0,0 +1,127 @@
+import React, { useCallback, useEffect, useRef, useState } from 'react'
+import { useDispatch, useSelector } from 'react-redux'
+import { bindActionCreators } from 'redux'
+
+import { BSPairType } from '@core/types'
+import BaseError from 'components/BuySell/Error'
+import { actions, selectors } from 'data'
+import { CARD_ERROR_CODE, FORMS_BS_BILLING_ADDRESS } from 'data/components/buySell/model'
+import { BSBillingAddressFormValuesType } from 'data/types'
+import { useRemote } from 'hooks'
+
+import { Loading, LoadingTextEnum } from '../../components'
+import { ADD_CARD_ERROR, IFRAME_ACTION } from '../AddCardCheckoutDotCom/model'
+import Success from '../AddCardCheckoutDotCom/template.success'
+import CardLoading from '../ThreeDSHandlerCheckoutDotCom/template.loading'
+import { getData } from './selectors'
+import { ReceiveMessageProps, VgsComponent } from './types'
+
+const AddCardVgs: VgsComponent = ({ handleClose }) => {
+  const dispatch = useDispatch()
+  const buySellActions = bindActionCreators(actions.components.buySell, dispatch)
+
+  const [isPolling, setPolling] = useState(false)
+  const [errorOnTokenization, setErrorOnTokenization] = useState(false)
+
+  const ref = useRef<HTMLIFrameElement>(null)
+
+  const { data, error, isLoading, isNotAsked } = useRemote(getData)
+  const { cardTokenId, vgsVaultId } = useSelector(selectors.components.buySell.getVgsAddCardInfo)
+  const cardSuccessRate = useSelector(selectors.components.buySell.getCardSuccessRate)
+  const formValues = useSelector(
+    selectors.form.getFormValues(FORMS_BS_BILLING_ADDRESS)
+  ) as BSBillingAddressFormValuesType
+
+  const handleReceivedPostMessage = useCallback(
```

Comment:
That would be a great candidate for refactoring into saga in my opinion 😀

Reply:
After doing some more research here I think you're right. Using eventChannels https://redux-saga.js.org/docs/advanced/Channels/#using-the-eventchannel-factory-to-connect-to-external-events we could create a saga that listens for the postMessage events which would dispatch the other events that are already in this `handleReceivedPostMessage` function. I'm going to create a ticket for myself to do this, but don't want to hold up this merge as I need to get backend testing this in staging so I can sort out some CORS errors with the.